lib/vmdb/plugins.rb

```diff
@@ -1,78 +1,85 @@
+require 'singleton'
+
 module Vmdb
   class Plugins
     include Singleton
+    include Enumerable
 
-    attr_reader :registered_ansible_content
-    attr_reader :registered_automate_domains
-
-    def initialize
-      @registered_automate_domains = []
-      @registered_provider_plugin_map = {}
-      @registered_ansible_content = []
-      @vmdb_plugins = []
+    def self.method_missing(m, *args, &block)
+      instance.respond_to?(m) ? instance.send(m, *args, &block) : super
```

Comment:
So, this is my rational for why I think we should can this and just include `Enumerable` in the class context, but that is all.  I am sure this is an "opinionated opinion" (redundant choice of words, but rolling with it) that you are free to ignore/contest/claim it is wrong and stupid, but think it is at least worth saying.

* * *

So I would much prefer putting `Enumerable` on the interface that is actually using it (the class), instead of applying it to the `.instance`, since thatcurrently isn't the public interface that is suggested to be used.

The only cases were we are directly using `Enumerable` functions (without having them delegated to via `method_missing` here) is inside of the methods defined in this class.  Frankly, it is not that much more work to write `all.each` or `all.flat_map`, and is probably easier for the next developer stumbling across this code to understand that `all` is the thing we are iterating on in those methods.

My suggestion [here](https://github.com/ManageIQ/manageiq/pull/17643/files#r198317057) should work, and wouldn't require someone figuring out what this `method_missing` is trying to do.  It also encourages the correct usage of calling `.each` and doesn't allow `.instance.each` to be valid.

Reply:
Whether you plan on keeping `.method_missing` or switching to my suggestion, I think it wouldn't hurt to have some documentation in comments explain how this class is supposed to be accessed and used.  Explaining things like how `.each` and such works, and maybe when the `.ansible_content`, `.load_inflections`, `.init`, and similar might be and are used.

Reply:
So, the reason I kept the .instance was for testing purposes with respect to caching.  When one uses class methods, and especially methods that cache values directly, then the probability of test contamination increases.  However, leaving the `.instance` allows the test author to stub out that one method with a mock or an independent Vmdb::Plugins.new, and since the caching lives inside that singleton instance, it will go away when the stub goes away.  Whether that will actually be useful or not it definitely debatable.  I'm honestly kind of experimenting a bit here to see if this way of structuring the code (Singleton+class-level forwarding vs straight-class-methods) is better or not from all perspectives.  I do know that I've had troubles in the past with testing straight-class-method things (e.g. Vmdb::Settings).

Reply:
Oh, I wasn't suggesting getting rid of the `include Singleton`, for what it is worth, just always defining the methods that actually use it on the class, instead of effectively duplicating them, and masking there class definition behind `Vmdb::Plugins.method_missing` (class level).

I totally understand using an `instance` to allow you to control state in tests and whatnot, and might be a better idea than what I suggested here:

https://github.com/ManageIQ/manageiq/pull/17643#issuecomment-400819800

So again, not contesting against using `Singleton` at all.  Simply the `method_missing` portion.

Reply:
I think I can also make instance a private method, which would solve your concern.

Reply:
Turns out I still need the `.method_missing` if I want to forward all the methods, and I like having the .each on the instance in case we want to test that in isolation.  I decided to just go with making .instance private, which we don't want people to use anyway, so it works to solve the higher-level problem of having 2 interfaces, one of which is forbidden.

Reply:
> Turns out I still need the `.method_missing`...

You could... you know... just forward the methods manually...

```ruby
module Vmdb
  class Plugins
    class << self
      include Enumerable
      
      def ansible_content; instance.ansible_content; end
      def automate_domains; instance.automate_domains; end
      # ...
    end
  end
end
```

* * * 

Bottom line:  I am not going to push for this any further since you are clearly going to stick with what you have, and we should just merge as it is.  But I will always think that `.method_missing` is a code smell and should be avoided at all costs, especially when the methods you are deferring to are a known quantity (like here).